### PR TITLE
RDFPFN792026: recognize guarded loading finalizers

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

React Doctor reports guarded finally resets as success-only loading resets even when a locally unique async-operation token or cleanup-backed mounted ref proves that the finalizer is safe.

## What changed

- recognize locally unique ownership tokens that are synchronously claimed before the risky await
- require ownership claims to precede `setLoading(true)` and reject unclaimed ref snapshots
- accept snapshot guards only when backed by a synchronous single-flight claim and paired release
- scan ownership writes across the complete enclosing lexical function scope while keeping separate sequence generators independent
- recognize committed effect invalidations only when paired with an unconditional reset of the same loading setter
- recognize cleanup-backed mounted refs while rejecting render-phase identities, mutable external request fields, branch-only resets, sibling writes, and stale ordered guards
- add focused adversarial tests, eight fuzz regression/true-positive fixtures, and a patch changeset

## ReactBench replay

The same 30 diagnostic units now partition as 7 proven false positives suppressed and 23 retained true positives reported. Bugbot exposed that one prior false-positive adjudication already had `setRefreshPending(true)` before its ownership token claim; that unit is now correctly reclassified as a true positive. All other 29 outcomes are unchanged.

Artifact: /tmp/reactbench-no-loading-finalizer-claim-order-replay-strict-7-23-5e03f5931.json
SHA-256: 2a17839263b53a9a3641815a16417719e834e138c638605a361c55d821385174

## Validation

- nr format
- nr lint (passes with repository baseline warnings)
- nr typecheck
- nr format:check
- nr test: all 15 tasks passed
- nr smoke:json-report
- nr -C packages/oxlint-plugin-react-doctor test src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.test.ts: 164 passed
- nr -C packages/oxlint-plugin-react-doctor test: 24,809 passed, 201 skipped
- nr -C packages/fuzz test: 192 passed, 782 skipped
- FUZZ_RULE=no-loading-flag-reset-outside-finally FUZZ_STRICT=1 FUZZ_ITERATIONS=500 nr fuzz
- nlx react-doctor@latest --verbose --scope changed: 100/100, no issues
- truffler post-change dedup searches: no competing helper matched

Local RDE is currently blocked before diagnostic ingestion because the latest eval checkout expects JSON schemaVersion 1 while current React Doctor emits schemaVersion 3.

---

> [!NOTE]
> **Medium Risk**
> Changes static analysis heuristics for a widely used lint rule; behavior shifts on edge cases (claim order, ordered guards, effect pairing) but is heavily test- and fuzz-backed.
> 
> **Overview**
> Tightens **`no-loading-flag-reset-outside-finally`** so conditional **`finally`** loading resets are not reported when the guard is provably tied to the current async work or a cleanup-backed mounted ref.
> 
> The rule now walks **`if` / `&&` / early-return** guard structure around the reset and accepts only **lifecycle guards** (effect cleanup–backed `mountedRef`, including sibling effects) and **claimed-ownership** checks (`ref.current === token`, request-id equality, ordered `>=` latest-start patterns). Ownership must be **synchronously claimed before** `setLoading(true)`; ref snapshots, claims after loading, stale `<=` ordered guards, outer-scope owner writes, and unpaired effect invalidation still **flag**.
> 
> Guarded finalizers that run **before a later risky `await`** outside the `try` are treated as safe. Implementation adds ownership-claim analysis across the enclosing lexical scope, single-flight snapshot exceptions, and pairs effect invalidation with an unconditional reset of the same setter.
> 
> Ships a **patch changeset**, **eight fuzz corpus fixtures**, and a large expansion of rule unit tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5e03f593179fc76a75e642a3ccca75002aab1471. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 1]
```diff
@@ -1673,10 +1767,250 @@ const isInsideTryFinalizer = (
...
+  if (claimIndex === -1 || truthyIndex === -1) return false;
+  if (claimIndex === truthyIndex) return true;
+  const firstIndex = Math.min(claimIndex, truthyIndex);
+  const lastIndex = Math.max(claimIndex, truthyIndex);
+  return claimEntry.block.body
+    .slice(firstIndex + 1, lastIndex)
+    .every(
+      (statement) =>
+        !subtreeHasAbruptSynchronousOperation(statement as EsTreeNode, functionNode, context),
+    );
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts:1831]: Fixed in e209abc1b3f1f8f565d68f902d064948ec5966d0. `claimPrecedesTruthySet` now requires the ownership claim to be a strictly earlier direct statement than `setLoading(true)`; same-statement claims are conservatively rejected because source order alone does not prove evaluation order. Added focused regressions for a later ref assignment, a later monotonic token claim, and a same-statement RHS setter, plus a deterministic true-positive fuzz fixture. Focused 158-rule tests, the full 24,803-test plugin suite, full repository tests, fuzz smoke, and strict fuzz 500 all pass. This also exposed that one prior ReactBench FP adjudication was this exact pattern, so the corrected 30-unit replay is 7 suppressed / 23 reported with the other 29 outcomes unchanged.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 2]
```diff
@@ -1673,10 +1767,247 @@ const isInsideTryFinalizer = (
...
+    serializeReferenceKey({ node: tokenInitializer.argument, scopes: context.scopes }) === refKey
+  ) {
+    candidates.push(tokenInitializer);
+  }
+  if (
+    tokenInitializer &&
+    getReactRefCurrent(tokenInitializer, context) &&
+    serializeReferenceKey({ node: tokenInitializer, scopes: context.scopes }) === refKey
+  ) {
+    candidates.push(tokenInitializer);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts]: Fixed in 30870c58906eae61831f46841f8af89aea86e710. A plain `const token = ref.current` snapshot is no longer an ownership candidate. Snapshot-based finalizers are accepted only when a separate synchronous single-flight claim is proven before the snapshot/setter (`if (lock.current) return` followed by `lock.current = true`), the same guarded finalizer releases that lock beside the loading reset, and no unpaired lock writes exist. Added focused report/accept regressions and a deterministic true-positive fuzz fixture. Focused 160 tests, all 24,805 plugin tests, full repository tests, lint, typecheck, format, strict fuzz 500, JSON smoke, and React Doctor 100/100 pass. The head-bound 30-unit replay remains exact at 7 suppressed / 23 reported with all 26 groups matching and no parse errors (artifact SHA-256 `2da39a3704baa563d1f0429b9d8323beb8e3408ea7cca5207dc44fc7cd5a3f7b`).

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 3]
```diff
@@ -1673,10 +1767,407 @@ const isInsideTryFinalizer = (
...
+      if (
+        writeTarget &&
+        serializeReferenceKey({ node: writeTarget, scopes: context.scopes }) === generationKey &&
+        !isEffectInvalidationPairedWithReset(candidate, truthySets, context)
+      ) {
+        didFindOtherGenerationWrite = true;
+      }
+    });
+    if (didFindOtherGenerationWrite) return null;
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts:2115]: Fixed in 1114399b148b272bbe8de289290f126c38ac3b28. Other writes to an incremented generation ref now invalidate the proof only when that same ref participates in the finalizer comparison. When the comparison uses a separately assigned owner ref, additional sequence increments only reserve unique tokens and do not revoke ownership. Added a focused zero-diagnostic regression and deterministic passing fuzz fixture. Focused 162 tests, all 24,807 plugin tests, full repository tests, lint, typecheck, format, strict fuzz 500, JSON smoke, and React Doctor 100/100 pass. The head-bound 30-unit replay remains exact at 7 suppressed / 23 reported with all 26 groups matching and no parse errors (artifact SHA-256 `90c74f9b65c7207ca94c5a615d1084904d033df581425f696a085b8f050b3b54`).

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 4]
```diff
@@ -1673,10 +1767,407 @@ const isInsideTryFinalizer = (
...
+        : null;
+    if (
+      writeTarget &&
+      serializeReferenceKey({ node: writeTarget, scopes: context.scopes }) === refKey &&
+      !isEffectInvalidationPairedWithReset(candidate, truthySets, context)
+    ) {
+      didFindOtherWrite = true;
+    }
+  });
+  return didFindOtherWrite ? null : claim;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts:2148]: Fixed in 1114399b148b272bbe8de289290f126c38ac3b28. Ownership-write scans now continue through every enclosing function and inspect the outermost lexical owner, so an async loader nested in an effect sees component-level writes to the captured ref. Added a focused reporting regression and deterministic true-positive fuzz fixture for an outer cancel handler. Focused 162 tests, all 24,807 plugin tests, full repository tests, lint, typecheck, format, strict fuzz 500, JSON smoke, and React Doctor 100/100 pass. The head-bound 30-unit replay remains exact at 7 suppressed / 23 reported with all 26 groups matching and no parse errors (artifact SHA-256 `90c74f9b65c7207ca94c5a615d1084904d033df581425f696a085b8f050b3b54`).

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 5]
```diff
@@ -1919,17 +2580,32 @@ const analyzeFunction = (functionNode: EsTreeNode, context: RuleContext): void =
...
+        firstRiskyAwait &&
+        isFinalizerResetProvablyGuarded(
+          call.node,
+          protectingTry,
+          functionNode,
+          truthySets,
+          firstRiskyAwait,
+          context,
+        )
+      );
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts:2606]: I could not reproduce this as a defect. The guard proof runs for this shape even when the protected `try` has no rejecting `await`: focused coverage now verifies both an ownership guard and a lifecycle guard before a later risky `await`, and both produce zero diagnostics. I also added the ownership case to the fuzz regression corpus. No rule-logic change was needed for this finding; the regression evidence is in `5e03f5931`.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 6]
```diff
@@ -1673,10 +1767,409 @@ const isInsideTryFinalizer = (
...
+    );
+    if (singleFlightClaim) candidates.push(singleFlightClaim);
+  }
+  if (tokenInitializer && isNodeOfType(tokenInitializer, "UpdateExpression")) {
+    const generationKey = serializeReferenceKey({
+      node: tokenInitializer.argument,
+      scopes: context.scopes,
+    });
+    if (!generationKey) return null;
+    if (generationKey === refKey) {
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts]: Fixed defensively in `5e03f5931`. A missing generation key no longer returns from `findOwnershipClaim`; it skips only that generation-ref candidate so later assignment-based ownership claims remain reachable. Focused tests, strict fuzzing, and the exact replay are green.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts [hunk 7]
```diff
@@ -1138,6 +1138,100 @@ const isCancellationGuardTest = (test: EsTreeNode): boolean => {
...
+    }
+    return null;
+  }
+  if (stripped.operator === "<=" && leftRef && rightToken) {
+    return { refCurrent: leftRef, token: rightToken, mode: "owns", isOrdered: true };
+  }
+  if (stripped.operator === ">=" && rightRef && leftToken) {
+    return { refCurrent: rightRef, token: leftToken, mode: "owns", isOrdered: true };
+  }
+  return null;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-loading-flag-reset-outside-finally.ts:1232]: These two comparisons are not equivalent ownership guards and accepting them would create a false negative. If the latest started request is `2` and a stale request is `1`, both `requestId <= latestStartedRef.current` (`1 <= 2`) and `latestStartedRef.current >= requestId` (`2 >= 1`) remain true, allowing the stale operation to clear the newer operation's loading state. `5e03f5931` adds focused and fuzz true-positive coverage for these stale orientations; the safe inverse/equality forms remain accepted.
